File: app/src/main/java/com/example/hitchapp/models/Ride.java
package com.example.hitchapp.models;

import org.json.JSONArray;
import org.parceler.Parcel;

import com.parse.ParseClassName;
import com.parse.ParseObject;
import com.parse.ParseUser;

@Parcel(analyze={Ride.class})
@ParseClassName("Ride")
public class Ride extends ParseObject{

    public static final String KEY_DRIVER = "driver";
    public static final String KEY_FROM = "from";
    public static final String KEY_TO = "to";
    public static final String KEY_DEPARTURE_TIME = "departureTime";
    public static final String KEY_PARTICIPANTS = "participants";
    public static final String KEY_PRICE = "price";
    public static final String KEY_PRICE_PER_PARTICIPANT = "pricePerParticipant";
    public static final String KEY_SEATS_AVAILABLE = "seatsAvailable";
    public static final String KEY_DEPARTURE_DATE = "departureDate";
    public static final String KEY_MESSAGES = "messages";
    public static final String KEY_REQUESTS = "requests";


    public ParseUser getDriver(){ return getParseUser(KEY_DRIVER); }

    public void setDriver(ParseUser user){
        put(KEY_DRIVER, user);
    }

    public JSONArray getParticipants(){return getJSONArray(KEY_PARTICIPANTS);}

    public void setParticipants(JSONArray array){put(KEY_PARTICIPANTS,array);}

    public JSONArray getRequests(){return getJSONArray(KEY_REQUESTS);}

    public void setRequests(JSONArray array){put(KEY_REQUESTS,array);}

    public String getFrom(){return getString(KEY_FROM);}

    public void setFrom(String from){put(KEY_FROM, from);}

    public String getTo(){return getString(KEY_TO);}

    public void setTo(String to){put(KEY_TO, to);}

    public String getDepartureTime(){return getString(KEY_DEPARTURE_TIME);}

    public void setDepartureTime(String time){put(KEY_DEPARTURE_TIME, time);}

    public String getDepartureDate(){return getString(KEY_DEPARTURE_DATE);}

    public void setDepartureDate(String time){put(KEY_DEPARTURE_DATE, time);}

    public int getPrice(){return (int) getNumber(KEY_PRICE);}

    public void setPrice(int price){put(KEY_PRICE, price);}

    public boolean getPricePerParticipant(){return getBoolean(KEY_PRICE_PER_PARTICIPANT);}

    public void setPricePerParticipant(boolean perParticipant){put(KEY_PRICE_PER_PARTICIPANT, perParticipant);}

    public int getSeatsAvailable(){return (int) getNumber(KEY_SEATS_AVAILABLE);}

    public void setSeatsAvailable(int seatsAvailable){put(KEY_SEATS_AVAILABLE, seatsAvailable);}

    public JSONArray getMessages() {return getJSONArray(KEY_MESSAGES);}

    public void setMessages(JSONArray messages){put(KEY_MESSAGES, messages);}
}

File: app/src/main/java/com/example/hitchapp/models/Ride.kt
package com.example.hitchapp.models

import com.parse.ParseClassName
import com.parse.ParseGeoPoint
import com.parse.ParseObject
import com.parse.ParseUser
import org.json.JSONArray
import org.parceler.Parcel
import java.util.*

@Parcel(analyze = [Ride::class])
@ParseClassName("Ride")
class Ride : ParseObject() {
    var driver: ParseUser?
        get() = getParseUser(KEY_DRIVER)
        set(user) {
            user?.let { put(KEY_DRIVER, it) }
        }

    var participants: JSONArray?
        get() = getJSONArray(KEY_PARTICIPANTS)
        set(array) {
            array?.let { put(KEY_PARTICIPANTS, it) }
        }

    var requests: JSONArray?
        get() = getJSONArray(KEY_REQUESTS)
        set(array) {
            array?.let { put(KEY_REQUESTS, it) }
        }

    var from: String?
        get() = getString(KEY_FROM)
        set(from) {
            from?.let { put(KEY_FROM, it) }
        }

    var to: String?
        get() = getString(KEY_TO)
        set(to) {
            to?.let { put(KEY_TO, it) }
        }

    var departureTime: String?
        get() = getString(KEY_DEPARTURE_TIME)
        set(time) {
            time?.let { put(KEY_DEPARTURE_TIME, it) }
        }

    var state: String?
        get() = getString(KEY_STATE)
        set(state) {
            state?.let { put(KEY_STATE, it) }
        }

    var fromLocation: ParseGeoPoint?
        get() = getParseGeoPoint(KEY_FROM_LOCATION)
        set(fromLocation) {
            fromLocation?.let { put(KEY_FROM_LOCATION, it) }
        }


    var departureDate: Date?
        get() = getDate(KEY_DEPARTURE_DATE)
        set(time) {
            time?.let { put(KEY_DEPARTURE_DATE, it) }
        }

    var price: Int
        get() = getNumber(KEY_PRICE) as Int
        set(price) {
            put(KEY_PRICE, price)
        }

    var pricePerParticipant: Boolean
        get() = getBoolean(KEY_PRICE_PER_PARTICIPANT)
        set(perParticipant) {
            put(KEY_PRICE_PER_PARTICIPANT, perParticipant)
        }

    var seatsAvailable: Int
        get() = getNumber(KEY_SEATS_AVAILABLE) as Int
        set(seatsAvailable) {
            put(KEY_SEATS_AVAILABLE, seatsAvailable)
        }
    companion object {
        const val KEY_DRIVER = "driver"
        const val KEY_FROM = "from"
        const val KEY_TO = "to"
        const val KEY_DEPARTURE_TIME = "departureTime"
        const val KEY_PARTICIPANTS = "participants"
        const val KEY_PRICE = "price"
        const val KEY_PRICE_PER_PARTICIPANT = "pricePerParticipant"
        const val KEY_SEATS_AVAILABLE = "seatsAvailable"
        const val KEY_DEPARTURE_DATE = "departureDate"
        const val KEY_REQUESTS = "requests"
        const val KEY_FROM_LOCATION = "fromLocation"
        const val KEY_STATE = "state"
    }
}